hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java

```diff
@@ -3409,19 +3464,55 @@ public ShortCircuitRegistry getShortCircuitRegistry() {
   public void checkDiskError() throws IOException {
     Set<FsVolumeSpi> unhealthyVolumes;
     try {
-      unhealthyVolumes = volumeChecker.checkAllVolumes(data);
+      // check all volume
+      unhealthyVolumes = volumeChecker.checkAllVolumes(allData);
       lastDiskErrorCheck = Time.monotonicNow();
     } catch (InterruptedException e) {
-      LOG.error("Interruped while running disk check", e);
+      LOG.error("Interrupted while running disk check", e);
       throw new IOException("Interrupted while running disk check", e);
     }
 
     if (unhealthyVolumes.size() > 0) {
+      if (errorDisk == null) {
+        errorDisk = new ArrayList<>();
+      }
+      List<StorageLocation> tmpDisk = Lists.newArrayList(errorDisk);
+      errorDisk = null;
+      for (FsVolumeSpi vol : unhealthyVolumes) {
+        LOG.info("Add error disk " + vol.getStorageLocation()
+            + " to errorDisk - " + vol.getStorageLocation());
+        errorDisk.add(vol.getStorageLocation());
```

Comment:
it will throw NPE since set `errorDisk=null` before loop.

Reply:
`errorDisk=null` should be `errorDisk.clear()`